### PR TITLE
Optimize block DAG

### PR DESCRIPTION
**Pull request by lochjin**

### CODE REVIEW

File: core/blockdag/blockdag.go [hunk 1]
```diff
@@ -351,10 +379,35 @@ func (bd *BlockDAG) GetBlock(h *hash.Hash) IBlock {
 // Acquire one block by hash
 // Be careful, this is inefficient and cannot be called frequently
 func (bd *BlockDAG) getBlock(h *hash.Hash) IBlock {
+	return bd.getBlockAdvanced(h, nil)
+}
+
+func (bd *BlockDAG) getBlockAdvanced(h *hash.Hash, tx database.Tx) IBlock {
```
Comment (dindinw) [inline core/blockdag/blockdag.go:385]: - looks `getBlockAdvanced` only called by `getblock` , it means `tx != nil` clause never executed
- the `if-else` clause surrounded by the duplicated logic of calling of `DBGetBlockIdByHash`, its better to refactor the duplicated logic itself into a func alone and eliminate the requisite for the `if-else` clause.

File: core/blockdag/blockdag_test.go [hunk 2]
```diff
@@ -308,15 +306,6 @@ func CalcBlockWeight(blocks int64, blockhash *hash.Hash, state byte) int64 {
 	return 1
```
Comment (dindinw) [inline core/blockdag/blockdag_test.go:306]: I think the mock logic in `CalcBlockWeight` might need to improve even in the unit test. the subsidy calculation is too important to ignore in the blockdag consensus test.

**Comment by dindinw** [comment]
@jamesvan2019 the `TestWsNotify ` test is still failed for the PR
https://travis-ci.com/github/Qitmeer/qitmeer/builds/222793934#L693

File: core/blockdag/blockdag.go (hunk 1 above)
Comment (lochjin) [reply core/blockdag/blockdag.go]: 62dda74

**Review by dindinw — APPROVED** [review]

**Comment by lochjin** [comment]
> @jamesvan2019 the `TestWsNotify ` test is still failed for the PR
> https://travis-ci.com/github/Qitmeer/qitmeer/builds/222793934#L693

https://github.com/Qitmeer/qitmeer/pull/476

**Pull request merged**